File: frontend/src/utils/youtube.js
export function extractVideoId(url) {

  if (!url) {
    return "";
  }

  const patterns = [

    /youtube\.com\/watch\?v=([^&]+)/,

    /youtu\.be\/([^?]+)/

  ];

  for (
    const pattern
    of patterns
  ) {

    const match =
      url.match(pattern);

    if (match) {

      return match[1];
    }
  }

  return url;
}

export function getThumbnailUrl(
  videoId
) {

  if (!videoId) {
    return "";
  }

  return (
    `https://img.youtube.com/vi/${videoId}/hqdefault.jpg`
  );
}

File: backend/app/utils/youtube.py
import re

def extract_video_id(url):
    if not url:
        return ""
    # Regular expression to match YouTube video IDs
    patterns = [
        r"youtube\.com/watch\?v=([^&]+)",
        r"youtu\.be/([^?]+)"
    ]

    for pattern in patterns:
        match = re.search(pattern, url)
        if match:
            return match.group(1)
    return url

